hnn_core/cell.py

```diff
@@ -293,42 +307,49 @@ def move_to_pos(self):
     # 2. a list needs to be created with a Dipole (Point Process) in each
     #    section at position 1
     # In Cell() and not Pyr() for future possibilities
-    def insert_dipole(self, yscale):
+    def insert_dipole(self, p_secs, dipole_sec):
         """Insert dipole into each section of this cell.
 
         Parameters
         ----------
-        yscale : dict
-            Dictionary of length scales to calculate dipole without
-            3d shape.
+        p_secs : nested dict
+            Dictionary with keys as section name.
+            p_secs[sec_name] is a dictionary with keys
+            L, diam, Ra, cm, syns and mech.
+            syns is a list specifying the synapses at that section.
+            The properties of syn are specified in p_syn.
+            mech is a dict with keys as the mechanism names. The
+            values are dictionaries with properties of the mechanism.
+        dipole_sec : str
+            The name of the section along which dipole is to
+            be projected.
         """
         self.dpl_vec = h.Vector(1)
         self.dpl_ref = self.dpl_vec._ref_x[0]
+        cos_thetas = _get_cos_theta(p_secs, 'apical_trunk')
 
-        # dends must have already been created!!
-        # it's easier to use wholetree here, this includes soma
-        sec_list = list(self.sections.values())
-        for sect in sec_list:
-            sect.insert('dipole')
-        # Dipole is defined in dipole_pp.mod
-        self.dipole_pp = [h.Dipole(1, sec=sect) for sect in sec_list]
         # setting pointers and ztan values
-        for sect, dpp in zip(sec_list, self.dipole_pp):
+        for sect_name in p_secs:
+            sect = self.sections[sect_name]
+            sect.insert('dipole')
+
+            dpp = h.Dipole(1, sec=sect)  # defined in dipole_pp.mod
+            self.dipole_pp.append(dpp)
             dpp.ri = h.ri(1, sec=sect)  # assign internal resistance
             # sets pointers in dipole mod file to the correct locations
             dpp._ref_pv = sect(0.99)._ref_v
             dpp._ref_Qtotal = self.dpl_ref
             # gives INTERNAL segments of the section, non-endpoints
             # creating this because need multiple values simultaneously
             pos_all = np.array([seg.x for seg in sect.allseg()])
-            # diff in yvals, scaled against the pos np.array. y_long as
-            # in longitudinal
-            sect_name = sect.name().split('_', 1)[1]
-            y_scale = (yscale[sect_name] * sect.L) * pos_all
-            # y_long = (h.y3d(1, sec=sect) - h.y3d(0, sec=sect)) * pos
+            pos_proj = sect.L * cos_thetas[sect_name] * pos_all
```

Comment:
this scaling is super mysterious to me. We're doing:

diff(section length * position vector of segment)
= section length * segment lengths

Reply:
Yeah I was trying to make sense of that too... I think it's very ugly shorthand. The L and theta are constant, so the diff is effectively on the position, which is in units of section length (`[0, 1]`). This may have something to do with calculating the _dipole moment_ of current, but it all depends on what `Dipole.mod` and `Dipole_pp.mod` are actually doing... I'm digging into the details of HOC (trying to figure out my LFP-woes). I have plans to visit the "Dipole"-calculations, too, hoping this part of our code will make more sense!

Reply:
oh I get it now. `sect.L * np.diff(pos_all)` is the fractional length (in units of um) of that segment. Multiply that by the current in the segment and you get the dipole moment of that segment

Reply:
Indeed, but that's what I mean by having to understand the MOD-files better. Because then they must not be calculating the dipole moment, right? But rather the primary current density...

Reply:
I think you're right. I'm guessing `Qsum` is the the sum of all radial current entering the dendrite at a given point in a nrn section?